src/persist/src/indexed/trace.rs

```diff
@@ -121,12 +177,98 @@ impl<K: Data, V: Data> BlobTrace<K, V> {
     ) -> Result<TraceSnapshot<K, V>, Error> {
         let ts_upper = self.ts_upper();
         let mut updates = Vec::with_capacity(self.batches.len());
-        for (_, key) in self.batches.iter() {
-            updates.push(blob.get_trace_batch(key)?);
+        for meta in self.batches.iter() {
+            updates.push(blob.get_trace_batch(&meta.key)?);
         }
         Ok(TraceSnapshot { ts_upper, updates })
     }
 
+    /// Merge two batches into one, forwarding all updates not beyond the current
+    /// `since` frontier to the `since` frontier.
+    fn merge<L: Blob>(
+        &mut self,
+        first: &BlobTraceBatchMeta,
+        second: &BlobTraceBatchMeta,
+        blob: &mut BlobCache<K, V, L>,
+    ) -> Result<BlobTraceBatchMeta, Error> {
+        if first.level != second.level {
+            return Err(format!(
```

Comment:
Should this really be an error? On one hand, it's unclear what the output level would be, but on the other, this feels to me more like a debug assertion in the caller of merge or something. I could be convinced

Edit: Thought about this more as I finished the review and I think it's fine to leave as-is, but something about this feels like a wart to me

Reply:
this could be a debug assert inside of merge?

Reply:
i placed debug asserts for "batches are the same level and contiguous"